app/Platform/Actions/UpdatePlayerGameMetricsAction.php

```diff
@@ -104,11 +154,156 @@ public function execute(PlayerGame $playerGame): void
+        if (!$this->playerBadgeExists($user->username, AwardType::GameBeaten, $game->id, $unlockMode)) {
+            $awardDate = Carbon::parse($this->calculateBeatenGameTimestamp($playerAchievements));
+
+            $badge = AddSiteAward(
+                $user->username,
+                AwardType::GameBeaten,
+                $game->id,
+                $unlockMode,
+                $awardDate,
+                displayOrder: 0
+            );
+
+            PlayerBadgeAwarded::dispatch($badge);
+            PlayerGameBeaten::dispatch($user, $game->id, $isBeatenHardcore);
+
+            if ($isBeatenHardcore && $awardDate->gte(Carbon::now()->subMinutes(10))) {
+                static_addnewhardcoregamebeaten($game->id, $user->username);
+            }
+        }
+    }
+
+    public function playerBadgeExists(string $username, int $awardType, int $data, ?int $dataExtra = null): bool
+    {
+         $badge = PlayerBadge::where('User', $username)
+            ->where('AwardType', $awardType)
+            ->where('AwardData', $data);
+
+        if ($dataExtra !== null) {
+            $badge->where('AwardDataExtra', $dataExtra);
+        }
+
+        return $badge->exists();
+    }
+
+    /**
+     * Beaten game awards are stored with an AwardDate that corresponds to when they
+     * unlocked the precise achievement that granted them the beaten status. This has
+     * to be calculated by on the rules that Progression and Win Condition achievements follow.
+     */
+    public function calculateBeatenGameTimestamp(mixed $userAchievements): string
+    {
+        $progressionAchievementsUnlocked = 0;
+        $latestProgressionDate = null;
+        $earliestWinConditionDate = null;
+
+        foreach ($userAchievements as $achievement) {
+            if ($achievement->type === AchievementType::Progression && $achievement->AchievementID) {
+                $progressionAchievementsUnlocked++;
+                // Keep track of the latest progression achievement date.
+                $latestProgressionDate = $latestProgressionDate === null || $achievement->Date > $latestProgressionDate
+                    ? $achievement->Date
+                    : $latestProgressionDate;
+            } elseif ($achievement->type === AchievementType::WinCondition && $achievement->AchievementID) {
+                // Keep track of the earliest win condition date.
+                $earliestWinConditionDate = $earliestWinConditionDate === null || $achievement->Date < $earliestWinConditionDate
+                    ? $achievement->Date
+                    : $earliestWinConditionDate;
+            }
+        }
 
-        $playerGame->update($metrics);
+        // Return the latest date between the progression and win condition achievements.
+        return $progressionAchievementsUnlocked > 0
+            ? ($latestProgressionDate ? max($latestProgressionDate, $earliestWinConditionDate) : $earliestWinConditionDate)
+            : $earliestWinConditionDate;
```

Comment:
In softcore mode, I unlocked a few achievements for a new user: TestMan123. I executed the following commands in order:

```
sail artisan ra:platform:player:unlock-achievement TestMan123 9,16,18,15,17,19
sail artisan ra:platform:player:unlock-achievement TestMan123 79434,79435,79436
sail artisan ra:platform:player:update-game-metrics TestMan123 1,14402
```

The `update-game-metrics` command crashed:

```
   TypeError

  App\Platform\Actions\UpdatePlayerGameMetricsAction::calculateBeatenGameTimestamp(): Return value must be of type string, null returned

  at app/Platform/Actions/UpdatePlayerGameMetricsAction.php:307
    303▕
    304▕         // Return the latest date between the progression and win condition achievements.
    305▕         return $progressionAchievementsUnlocked > 0
    306▕             ? ($latestProgressionDate ? max($latestProgressionDate, $earliestWinConditionDate) : $earliestWinConditionDate)
  ➜ 307▕             : $earliestWinConditionDate;
    308▕     }
    309▕ }
    310▕
```

Dragster (game 14402) does not have any typed achievements (no progression or win condition) in my local database.

Reply:
After running:

```
sail artisan ra:platform:player:update-game-metrics TestMan123 1
sail artisan ra:platform:game:update-metrics 1`
```

I see TestMan123 as having unlocked every achievement for game 1.

![Screenshot 2023-09-16 at 11 32 29 AM](https://github.com/RetroAchievements/RAWeb/assets/3984985/d6a708fc-bf2b-4ca7-bdd5-30970dfb9609)

Reply:
I can generate the same error with this command:
```
php artisan ra:platform:player:update-game-metrics MikeyG 11008
```
The game has progression achievements flagged, but the user has not unlocked any of the progression achievements.

Reply:
Also when resetting all achievements for the game:
![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/5a9e4a19-0dad-4b38-97fd-bb5451f9a26b)

Reply:
I assume it's related to this error, but unlocked achievements are not awarding contribution points or mastery badges.

Reply:
`calculateBeatenGameTimestamp()` wasn't fully refactored yet.
Please check for those errors again in latest.

Reply:
These issues now appear to be resolved for me.